Yago/step12.2

YagoAccenture opened the pull request:
Deletion of entity in domain layer, and created in the presentation layer


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
  - Introduced a new use case for fetching episode details and associated TMDB information.
  - Enhanced `DetailsActivity` to display episode ratings and images.
  - Added new UI components for episode rating and imagery in the layout.
  - New data classes created to represent episode details and season/episode relationships.

- **Bug Fixes**
  - Corrected typographical errors in class and method names related to TMDB functionality.

- **Improvements**
  - Updated data handling for episode details to include TMDB data, improving overall functionality and user experience.

- **Style**
  - Improved layout structure and visual alignment of UI components in the details screen.
  - Added a new string resource for displaying character ratings.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
Step 12.1 ADD:  Show the chapter rating

Commit:
Step 12.1 EDIT:
val tmdbRepository = TmdbReposirory(tmdbRemoteDataSource) -> TmdbRepository(tmdbRemoteDataSource)

Commit:
Step 12.1 EDIT:
Unification and renaming of the usecase, aesthetic modification of DetailsActivity, to display the new data correctly

Commit:
Step 12.1 EDIT:
Separate use cases to leave the code as before

Commit:
Step 12.1 ADD:
Use cases

Commit:
Step 12.1 EDIT:
Testcases unified: GetEpisodeTMDBUseCase and GetEpisodeUseCase

Commit:
Step 12.1 EDIT:
Rename of use case UnificatedUseCase -> GetEpisodeAndDetailsUseCase

Commit:
Step 12.1 EDIT:
Deletion of entity in domain layer, and created in the presentation layer

PabloGrelaM reviewed (changes requested):
Review GetEpisodeAndDetailsUseCase

Commit:
Step 12 EDIT:
Presentation layer entity corrected so that it does not use entities from another layer for its creation

Commit:
Step 12 EDIT:
Substitution of the Pair, with the expected entities

PabloGrelaM committed:
Modified coderabbit config.

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/domain/GetEpisodeAndDetailsUseCase.kt:
```diff
@@ -0,0 +1,52 @@
+    private val tmdbRepository: TmdbRepository
+) {
+    suspend operator fun invoke(episodeUrl: String): Result<EpisodeDetailsViewEntity> {
+        val episodeResult = repository.fetchEpisodeDetails(episodeUrl)
```
this repo shouldn't return a Result<T> it should return a T

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/domain/GetEpisodeAndDetailsUseCase.kt:
```diff
@@ -0,0 +1,52 @@
+        return Result.failure(Exception("Failed to load episode details"))
+    }
+
+    private fun extractSeasonAndEpisode(episodeCode: String?): SeasonAndEpisode {
```
This should return a nullable object, if the parse fails, a null should be returned.
Also change the substring for a REGEX.

YagoAccenture replied:
Solved in next commit

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/domain/GetEpisodeAndDetailsUseCase.kt:
```diff
@@ -0,0 +1,52 @@
+            val tmdbResult = tmdbRepository.getEpisodeTMDBInfo(seasonNumber, episodeNumber)
+
+            if (tmdbResult != null) {
+                val episodeDetailsViewEntity = mapToEpisodeDetailsViewEntity(
```
change this map to be consistent with the rest of the project

YagoAccenture replied:
Solved in next commit

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt:
```diff
@@ -23,18 +23,25 @@ class DetailsViewModel(
-            episodeResult.getOrNull()?.let {
-                view.displayFirstEpisodeDetails(it)
+            val result = getEpisodeAndDetailsUseCase(episodeUrl)
+            if (result.isSuccess) {
```
unsae result.onSuccess{}.onFailure{}

YagoAccenture replied:
Solved in next commit¿?

PabloGrelaM replied:
nope

YagoAccenture replied:
I think that now is solved in the next commit.

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt:
```diff
@@ -23,18 +23,25 @@ class DetailsViewModel(
 interface DetailsViewTranslator {
     fun displayCharacterDetails(character: CharacterEntity)
-    fun displayFirstEpisodeDetails(episode: EpisodeEntity)
+    fun displayFirstEpisodeDetails(episodeDetailsViewEntity: EpisodeDetailsViewEntity)
```
don't use two different methods with the same param and being called in the same method.
Split them to handle the option of one backend failing and the other working or make the assumption that both are going to work.

YagoAccenture replied:
Solved in next commit

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/presentation/SeasonAndEpisode.kt:
```diff
@@ -0,0 +1,6 @@
+package com.mstudio.superheroarch.presentation
+
+data class SeasonAndEpisode(
+    val seasonNumber: Int?,
```
those vals are not nullable

YagoAccenture replied:
Solved in next commit

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/ui/activity/DetailsActivity.kt:
```diff
@@ -66,6 +77,23 @@ class DetailsActivity : AppCompatActivity(), DetailsViewTranslator {
+        ratingTextView.text = "Rating: ${episodeDetailsViewEntity.rating}"
+
+        episodeDetailsViewEntity.imageUrl.let {
+            Picasso.get().load(it)
```
this let is not needed.

YagoAccenture replied:
Solved in next commit.

PabloGrelaM commented on app/src/main/res/layout/details_screen.xml:
```diff
@@ -203,7 +206,7 @@
                 android:layout_width="wrap_content"
                 android:layout_height="wrap_content"
-                android:layout_marginTop="8dp"
+                android:layout_marginTop="0dp"
```
if margin is 0, don't set any margin.

YagoAccenture replied:
Solved in next commit.

PabloGrelaM reviewed (changes requested):
After this PR, create another one creating Models for each layer.

Commit:
Step 12 EDIT:
Removed, let it not need, removed 0dp margin, and added a string for rating

Commit:
Step 12 EDIT:
Elimination of 2 nunable values ​​that cannot be nunable, and 2 methods with the same parameter called in the same method are no longer used

Commit:
Step 12 EDIT:
Mapper modified to be the same as the others in the app, the repo now returns a T and not a Result

Commit:
Merge branch 'yago/step12.2' of https://github.com/cmm-academy/SuperHeroArch into yago/step12.2

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/presentation/EpisodeDetailsViewEntity.kt:
```diff
@@ -0,0 +1,22 @@
+package com.mstudio.superheroarch.presentation
+
+data class EpisodeDetailsViewEntity(
+    val air_date: String,
```
air_date is not a valid naming

YagoAccenture replied:
Solved in next commit

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/repository/RickAndMortyRepository.kt:
```diff
@@ -29,12 +29,11 @@ class RickAndMortyRepository(
-            Result.failure(e)
+            remoteDataSource.getEpisode(episodeUrl)
+        } catch (e: IOException) {
+            throw e
```
catch + throw means nothing. You are catching the exception and throwing it again, this is useless. Remove the try catch.

YagoAccenture replied:
Solved in next commit

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/ui/activity/DetailsActivity.kt:
```diff
@@ -1,6 +1,7 @@
 package com.mstudio.superheroarch.ui.activity
 
 import android.os.Bundle
+import android.util.Log
```
unused import.

YagoAccenture replied:
Solved in next commit

Commit:
Step 12 EDIT:
Fixed comments errors such as:
unused dependencies, variable name change

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt:
```diff
@@ -23,18 +24,23 @@ class DetailsViewModel(
 
     private suspend fun fetchFirstEpisodeDetails(episodeUrl: String) {
         try {
```
You are doing the try catch in the usecase, that catch returns a Result.failure, so you don't need this try catch.

YagoAccenture replied:
Solved in next commit

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/presentation/EpisodeDetailsViewEntity.kt:
```diff
@@ -0,0 +1,22 @@
+)
+
+fun EpisodeDetailsViewEntity.mapToViewEntity(
+    air_date: String,
```
air_date is still not a valid naming.

YagoAccenture replied:
Solved in next comit

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/ui/activity/DetailsActivity.kt:
```diff
@@ -76,9 +86,24 @@ class DetailsActivity : AppCompatActivity(), DetailsViewTranslator {
+        episodeEntity: EpisodeEntity?,
+        episodeDetailsViewEntity: EpisodeDetailsViewEntity?
+    ) {
+        if (episodeDetailsViewEntity != null) {
```
this if should be in the viewmodel.

Commit:
Step 12 EDIT:
Rename airdate, removal of try catch in DetailsViewModel, if statement is now in the viewmodel

Commit:
Step 12 EDIT:
Warning solved
